### PR TITLE
Proposal: Make all main logic available in caddy package

### PR DESCRIPTION
**Pull request by captncraig**
One problem I have is creating a custom build with a custom set of plugins. The steps are straightforward, but a bit tricky:

1. Install caddyext
2. get main repo and all plugins
3. run caddyext
4. build caddy

When updates happen only some of this cycle needs to be repeated but problems are also created by caddyext making my caddy clone "dirty", and requiring more git tricks.

I propose a few simple changes to make dealing with custom build easier. The main server logic is currently available in the `caddy` package, but not the main logic, which essentially consists of options parsing and so forth. This does two things:

1. Provides a mechanism for directives to be added at runtime into the directives stack. This alone should be useful to anyone wanting to run caddy as a library.
2. Provide all main logic as `caddy.Main()`. the main package is now a one line call into `caddy.Main()`. 

This essentially allows me to make my own "fork" of caddy. I can make a tiny main that registers the directives I need and run main. I can also vendor code for both caddy and plugins together in my own repo, which is another problem made harder by the multi-repo plugin style caddy has.

cc @pedronasser @abiosoft

### CODE REVIEW

**Comment by captncraig** [comment]
https://github.com/captncraig/caddycustom is an example of what I can do with this. A single repo I can update and build, complete with vendored dependencies for all my plugins. I can make a docker image from this with almost no effort.

**Comment by mholt** [comment]
(The failing test on AppVeyor is a race condition in the test case; don't mind it.)

File: caddy/directives.go [hunk 1]
```diff
@@ -68,6 +68,20 @@ var directiveOrder = []directive{
 	{"browse", setup.Browse},
 }
 
+func RegisterDirective(name string, setup SetupFunc, after string) {
```
Comment (abiosoft) [inline caddy/directives.go]: As this is a public function, it needs documentation.

**Comment by abiosoft** [comment]
I had a discussion with Matt recently about exposing the ability to add directives at runtime. :+1:

**Comment by captncraig** [comment]
Added doc comment and rebased.

**Comment by mholt** [comment]
After looking it over, I get why RegisterDirective could be useful, but why move main into caddy?

**Comment by captncraig** [comment]
So I can write my own main package that includes the external plugins I want without rewriting caddy source code (including potentially private ones). This makes deploying my custom build as easy as `go get -u mypackage`. I can also vendor dependencies for caddy and plugins together in a uniform way, and I really can't think of a good way to accomplish that at all.

See https://github.com/captncraig/caddycustom/blob/master/main.go for how I do it now. I wouldn't expect everybody to build it this way, but IMO it is a nice option to have. Without this PR, I would essentially copy and paste a lot of your code to make sure I end up with a binary that handles flags and such the same as "real" caddy.

**Comment by mholt** [comment]
Ah. I see what you mean now.

This is not going to work unless main.go is in its own package. Right now, any program importing the caddy package would lose control of all signals and have flags added to its CLI.

**Comment by captncraig** [comment]
That's a good point too. Perhaps the flags can be registered inside the
main function so they don't leak?

**Comment by mholt** [comment]
Sounds scary, gotta make sure taking it out of init won't break anything...

**Comment by captncraig** [comment]
I moved flag declarations into main. Looks to me like they all parse fine, and `-h` still works. As long as they are declared before `flag.Parse()` I don't think it matters.

**Comment by mholt** [comment]
Cool. This seems like a good time to bring this up: https://github.com/spf13/caddy/commit/6a9a368c71966fb990f1c63a6b50681d10b03d58

There are definitely advantages to moving to cobra for CLI. Would it help your use case too?

**Comment by captncraig** [comment]
I am not opposed to cobra. It does make for nice commands. I don't think this pr prevents or helps that initiative significantly one way or another.

**Comment by mholt** [comment]
Right, but merging one (if the other repo had a PR opened, that is) will break the other.

Sorry, I still need a little more time on this. But I think it's looking pretty compelling now.

**Comment by captncraig** [comment]
I realize I do not actually need the main stuff to do what I want to do. If the registration stuff is present, I can simply add a new file to the main package (in my custom fork) that registers my plugins in an `init`

**Comment by captncraig** [comment]
I'll purge the main stuff from this pr.

**Comment by mholt** [comment]
That's a brilliant idea, and that way I don't have to worry about merge conflicts if I switch Caddy over to a cobra CLI app. ;)

**Comment by captncraig** [comment]
Replaced by #548

**Pull request closed without merging**